src/core/ParticleIterator.hpp

```diff
@@ -16,6 +16,10 @@ struct ParticleIterator : public boost::iterator_facade<
     }
   }
 
+  operator Particle *() const {
```

Comment:
How about a `get()` function (or `ptr()` if you want)?  If you don't think these suit you, then please at least make the conversion `explicit`.  Otherwise the iterator might decay to a `Particle*` in unexpected places.